course4/lecture4_en.md

```diff
@@ -0,0 +1,306 @@
+    Cons(_, tl) => contains_zero(tl)
+  }
+}
+```
+
+In this example, the first `Cons` uses the number 0 to match lists starting with 0, and the second `Cons` uses a wildcard and an identifier `tl` to match remaining lists while binding the sublist to the identifier `tl` for further processing. The value of the current list is discarded by the wildcard.
+
+Pattern matching for tuples and structures is just like for constructions. 
+
+```moonbit
+fn first(pair: (Int, Int)) -> Int {
+  match pair {
+    (first, second) => first
+  }
+}
+
+fn baby_name(info: PersonalInfo) -> Option[String] {
+  match info {
+    { age: 0, .. } => None
+    { name, age } => Some(name)
+  }
+}
+```
+
+Tuples' patterns are just like their definitions, enclosed in parentheses and separated by commas. Make sure the length of the matched tuple is correct. Structure patterns are enclosed in braces and separated by commas. We have additional pattern forms to make pattern matching more flexible:
+
+* Explicitly match some specific values, such as `age: 0` to match the data with specific values. 
+* Use another identifier to bind a field, such as `age: my_age`. This is useful when you don't want to use the field name as an identifier.
+* Omit remaining fields with `..` at the end.
+
+Here is another example for better understanding how to use nested patterns. The `zip` function combines two lists into a new list of pairs like a zipper. The length of the resulting list is the minimum of the lengths of the input lists. Given the lists [1, 2, 3] and [a, b, c, d], the zipped list would be (1, a) (2, b) (3, c).
+
+```moonbit
+fn zip(l1: List[Int], l2: List[Char]) -> List[(Int, Char)] {
+  match (l1, l2) {
+    (Cons(hd, tl), Cons(hd2, tl2)) => Cons((hd, hd2), zip(tl, tl2))
+    _ => Nil
+  }
+}
+```
+
+We define our function with pattern matching. Here, we match a pairs by constructing a tuple and then match the nested tuple pattern, effectively matching both lists simultaneously. If either of the input lists is empty, the result is an empty list. When both lists are non-empty, we get a non-empty result. The first item of the result is a tuple of the two values we take from the inputs, followed by the zipped result of the sublists of both lists. Note that the order of pattern matching is top-down. (If a wildcard is placed at the top, the subsequent patterns will never be matched, and the code will never run. The good news is that MoonBit can detect this and provide warnings. These warnings are advisory and won't prevent compilation, so it's crucial to pay attention to the issues panel in your IDE.)
+
+Lastly, pattern matching is not limited to `match`; it can also be used in data binding. In local definitions, we can use pattern matching expressions to bind corresponding substructures to identifiers. It's essential to note that if value comparison occurs at this point and the match fails, the program will encounter a runtime error, leading to program termination.
```

Comment:
Maybe an example here? Also what does 'if value comparison occurs' mean?

Reply:
The example of *let binding pattern*? (The zip function definition is above)

'if value comparison occurs' is directly translated from Chinese. I think what we want to express here is the difference between 'refutable/irrefutable' patterns. But I am not sure if they are standard terminologies in MoonBit.

Reply:
> The example of let binding pattern? (The zip function definition is above)

Yes

> 'if value comparison occurs' is directly translated from Chinese.

Ok, I see what you mean. Maybe we can just omit it and say 'if the match fails'